### PR TITLE
Add type hints to `tests/rest/client`

### PR DESCRIPTION
**Pull request by dklimpel**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: tests/rest/client/utils.py [hunk 1]
```diff
@@ -54,9 +55,9 @@ class RestHelper:
     REST action, which isn't the focus of the test.
     """
 
-    hs = attr.ib()
+    hs = attr.ib(type=HomeServer)
     site = attr.ib(type=Site)
-    auth_user_id = attr.ib()
```
Comment (DMRobertson) [inline tests/rest/client/utils.py:59]: We generally prefer using `auto_attribs=True` if possible, which allows us to use a slightly neater syntax in the class body. See for example

https://github.com/matrix-org/synapse/blob/5a6911598ad2d3dea96b9f8c1cffccd4f4840bf7/synapse/storage/databases/main/events_worker.py#L107-L110

File: tests/rest/client/utils.py [hunk 2]
```diff
@@ -145,7 +146,7 @@ def create_room_as(
 
     def invite(
         self,
-        room: Optional[str] = None,
+        room: str,
```
Comment (DMRobertson) [inline tests/rest/client/utils.py:149]: It doesn't make sense to invite to, leave, or ban from an unspecified room. Do you know why this annotation allowed an `Optional` room in the first place? Are/were there any tests taking advantage of this?

File: tests/rest/client/test_retention.py [hunk 3]
```diff
@@ -283,12 +290,15 @@ def test_state_policy(self):
...
-    def _test_retention(self, room_id, expected_code_for_first_event=200):
+    def _test_retention(
+        self, room_id: str, expected_code_for_first_event: int = 200
+    ) -> None:
         # Send a first event to the room. This is the event we'll want to be purged at the
         # end of the test.
         resp = self.helper.send(room_id=room_id, body="1", tok=self.token)
 
         first_event_id = resp.get("event_id")
+        assert first_event_id
```
Comment (DMRobertson) [inline tests/rest/client/test_retention.py:301]: I think we normally prefer `assert first_event_id is not None` when we're narrowing `Optional[T]` to `T`.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Thanks for this; I'm happy if mypy is happy. A few style notes though.

Otherwise,

File: tests/rest/client/utils.py (hunk 2 above)
Comment (dklimpel) [reply tests/rest/client/utils.py:149]: It calls `change_membership`.
There it is required and also needed for the path.

https://github.com/matrix-org/synapse/blob/2b0f1575a9145cabba92beea8969af2390d147eb/tests/rest/client/utils.py#L252-L283

Comment (dklimpel) [reply tests/rest/client/utils.py:149]: It was part of the inital Homeserver commit:
https://github.com/matrix-org/synapse/commit/4f475c7697722e946e39e42f38f3dd03a95d8765#diff-7d9372b9840620c86d8739fc6be631da0bcdcc4b41101b435dc4429ad34b6094

**Review by DMRobertson — APPROVED** [review]
Many thanks Dirk! LGTM. Once the unit tests have finished running, I'll merge this. Assuming sytest is still screwed for now.

**Pull request merged**